06x-new-kafka-roller.md

```diff
@@ -0,0 +1,350 @@
+
+Depending on the observed states, the roller will perform specific actions. Those actions should cause a subsequent observation to cause a state transition. This iterative process continues until each node's state aligns with the desired state.
+
+It will also introduce an algorithm that can restart brokers in parallel when safety conditions are not violated. These conditions guarantee Kafka producer availability and cause minimal impact on controllers and overall stability of clusters.
+
+### Node State
+When a new reconciliation starts up, a context object is created for each node to store the state and other useful information used by the roller. It will have the following fields:
+
+ - <i>nodeRef</i>: NodeRef object that contains Node ID.
+ - <i>currentNodeRole</i>: Currently assigned process roles for this node (e.g. controller, broker).
+ - <i>lastKnownState</i>: It contains the last known state of the node based on information collected from the abstracted sources (Kubernetes API, KafkaAgent and Kafka Admin API). The table below describes the possible states.
+ - <i>restartReason</i>: It is updated based on the current predicate logic from the `Reconciler`. For example, an update in the Kafka CR is detected.
+ - <i>numRestartAttempts</i>: The value is incremented each time the node has been attempted to restart.
+ - <i>numReconfigAttempts</i>: The value is incremented each time the node has been attempted to reconfigure.
+ - <i>numRetries</i>: The value is incremented each time the node cannot be restarted/reconfigured due to not meeting safety conditions (more on this later).
+ - <i>lastTransitionTime</i>: System.nanoTime of last observed state transition.
+
+ The following table illustrates the proposed node states and the possible transitions:
+ | State            | Description      | Possible transitions |
+ | :--------------- | :--------------- | :----------- |
+ | UNKNOWN               | The initial state when creating `Context` for a node. We expect to transition from this state fairly quickly after creating the context for nodes.  | `NOT_RUNNING` `NOT_READY` `RECOVERING` `SERVING` |
+ | NOT_RUNNING           | Node is not running (Kafka process is not running). This is determined via Kubernetes API, more details for it below. | `NOT_READY`, `NOT_RUNNING`. | `RESTARTED` `SERVING` | 
+ | NOT_READY             | Node is running but not ready to serve requests which is determined by Kubernetes readiness probe (broker state < 2 OR == 127 OR controller is not listening on port). | `RESTARTED` `SERVING` |
+ | RESTARTED             | After successful `kubectl delete pod`. | `NOT_RUNNING` `NOT_READY` `RECOVERING` `SERVING` |
+ | RECONFIGURED          | After successful Kafka node config update via Admin client. | `NOT_RUNNING` `NOT_READY` `RESTARTED` `SERVING` |
```

Comment:
The states don't necessarily align with Kubernetes states but I do see what you mean, as most states align with underlying systems. I thought It's useful to know that an action has been successfully taken on the node, so for example, the state transitions from `READY` -> `RESTARTED` -> `READY`. However, I guess the `numRestartAttempts`/`maxReconfigAttempts` indicates the action too so perhaps these states (`RESTARTED`/`RECONFIGURED`) are not needed.

Reply:
I think that if you need some transitional state for some technical reasons, would it be better to describe the uncertainty after the operation? E.g. `READY` -> `UNKNOWN` (after the restart) -> `???` (e.g. READY after the next check)?

That said, purely from the Kubernetes perspective:
* The roller deletes the Pod, it does not restart it. So it actually transitions it to `NOT_RUNNING` after it is deleted
* The StirmziPodSetController starts the pod again which is when it might move to `NOT_READY` and later to `READY`

But that reminds me ... there will be also `Terminating` phase I guess that might be there long enough for you to map it somewhere?

Reply:
> We want the states to also reflect the action taken on the pods. So that when we do the observation again, we wait for the state to change to NOT_READY, or READY kind of states.

I think it would be better having some additional fields in the node context taking track of restarted/reconfigured if you need them. But I think they are not actual states within an FSM.

Reply:
Wouldn't `RESTARTED` will mean somewhat similar to `UNKNOWN` in this case? It would only mean that  the pod was restarted and not it can can  be `Ready` or `notReady` in the next phase while at the same it also clarifies the action which is taken i.e(the pod is restarted)

Reply:
@tombentley @ShubhamRwt any thoughts on this? I think removing the action states and using the unknown state after the action is taken is ok too.